[Android] Fix image disappearing when leaving `ShellSection`

albyrock87 opened the pull request:
### Description of Change

We realized that Glide clears the target when the fragment moves to DESTROYED state, which happens for example while switching between `ShellSection`s.

Destroying the `Fragment` causes the `Target` to trigger an `onDestroy` and `onResourceCleared` methods, and that's where we store a new information regarding the target.

We have to reapply the `Source` property when the image is reattached to the window, only if it didn't failed.

This PR also adds better logging around `Glide` integration which can be enabled through some adb commands:

Base logging:
```
adb shell setprop log.tag.Glide VERBOSE
```

Additional logging:
```
adb shell setprop log.tag.CustomViewTarget VERBOSE
adb shell setprop log.tag.GlideRequest VERBOSE         
adb shell setprop log.tag.Engine VERBOSE 
adb shell setprop log.tag.HttpUrlFetcher VERBOSE 
```

This PR also adds a test case (without automation) to better ensure no regression when switching between images, including `null` and invalid URI.

### Issues Fixed

Fixes #14471

albyrock87 committed:
Don't let Glide clear the old image on Android ImageView unless we get an error

kubaflo committed:
Added a UI Test (#14471)

albyrock87 committed:
Use `DrawableImageViewTarget` instead of using our own implementation and just invoke C# callbacks on top of it

albyrock87 committed:
Use view tags to store additional information regarding Glide load status/resources

albyrock87 committed:
Address flickering in a different way by avoiding ImageView invalidation while loading

albyrock87 committed:
Added UI tests screenshots

albyrock87 committed:
Fix implementation and clarify with comments

jfversluis commented:
/azp run

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformInterop.java:
```diff
@@ -64,6 +71,16 @@
+    
+    public static boolean getGlideClearedTag(View view) {
+        Object tag = view.getTag(MAUI_GLIDE_CLEARED_TAG_ID);
+        return tag != null && (boolean)tag;
```
Does this need to check `instanceof` in case it is another type?
```suggestion
        return tag != null && tag instanceof boolean && (boolean)tag;
```

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/glide/MauiCustomViewTarget.java:
```diff
@@ -13,51 +14,90 @@
+            this.platformView.freeze();
+        }
+        
+		if (this.destroyed && !this.failed) {
+		    logV("onResourceCleared: setIsCleared: true");
+            PlatformInterop.setGlideClearedTag(this.view, true);
+        }
```
Formatting: something is extra indented here.

jonathanpeppers commented on src/Core/src/Handlers/Image/ImageHandler.Android.cs:
```diff
@@ -21,9 +20,14 @@ protected override ImageView CreatePlatformView()
+
 		protected override void DisconnectHandler(ImageView platformView)
 		{
-			base.DisconnectHandler(platformView);
```
Did we remove `base.DisconnectHandler()` on purpose?

albyrock87 replied:
Base method is empty and many other handlers do not call the base method.
Anyway, I'll bring it back just in case someone in the future wants to fix the situation in general.

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformInterop.java:
```diff
@@ -64,6 +71,16 @@
 import java.util.List;
 
 public class PlatformInterop {
+    private static final int MAUI_GLIDE_CLEARED_TAG_ID = R.id.maui_glide_cleared_tag;
+    
+    public static void setGlideClearedTag(View view, boolean isCleared) {
+        view.setTag(MAUI_GLIDE_CLEARED_TAG_ID, isCleared);
+    }
```
Instead of using `view.setTag()`, could we have a `boolean isCleared` on `PlatformAppCompatImageView` instead? Could it just be a regular field, with get/set methods? I'm not sure how often developers would try to use the tag for other data, but they can.

albyrock87 replied:
I used the tag simply because I wanted to provide this functionality also on other views.
But now I realize this only makes sense on Glide View Targets which happens to be only the image view.
So I agree with you, I'll move this to a property.

albyrock87 replied:
On a second thought there was another, more important, reason.
At the moment `ImageHandler` is exposing `ImageView` so customers may have overridden the handler and provide a different platform view (providing their subclass of `ImageView`).
For those customers we still want to fix #14471 while it is acceptable (I guess) to leave them with an eventual flicker.

jonathanpeppers commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/glide/MauiCustomViewTarget.java:
```diff
@@ -13,51 +14,90 @@
+import com.microsoft.maui.PlatformInterop;
 
 public class MauiCustomViewTarget extends CustomViewTarget<ImageView, Drawable> {
+    private static final String TAG = "MauiCustomViewTarget";
+    private static final boolean IS_VERBOSE_LOGGABLE = Log.isLoggable(TAG, Log.VERBOSE);
```
I like the logging you've added throughout, but should we have a single tag, `Log.isLoggable("Glide", Log.VERBOSE)` for now? And not create multiple tags? Maybe also create a `GlideLogging` class to centralize it?

albyrock87 committed:
Update src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformInterop.java

Co-authored-by: Jonathan Peppers <jonathan.peppers@gmail.com>

albyrock87 committed:
Addressing PR feedbacks

albyrock87 committed:
Simplify logging and freeze only if there's a drawable

albyrock87 committed:
Enable memory cache on `FontModel`

albyrock87 committed:
Improve testing page

albyrock87 committed:
Address requests from core team
- Remove FontModel cache fix
- Remove freeze strategy
- Don't commit `maui.aar`

mattleibow commented on src/Core/src/Handlers/Image/ImageHandler.Android.cs:
```diff
@@ -21,9 +21,15 @@ protected override ImageView CreatePlatformView()
 		protected override void DisconnectHandler(ImageView platformView)
 		{
 			base.DisconnectHandler(platformView);
+			platformView.ViewAttachedToWindow -= OnPlatformViewAttachedToWindow;
```
@PureWeen will this cause a leak because we never call disconnect?

mattleibow commented on src/Core/src/Handlers/Image/ImageHandler.Android.cs:
```diff
@@ -78,6 +84,14 @@ public override void PlatformArrange(Graphics.Rect frame)
+		
+		void OnPlatformViewAttachedToWindow(object? sender, View.ViewAttachedToWindowEventArgs e)
+		{
+			if (PlatformView is {} platformView && PlatformInterop.GetGlideClearedTag(platformView) && VirtualView is { } virtualView)
```
I think this should also have a platformView.IsAlive()... The GC loves to crash things by disposing.

mattleibow commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/glide/font/FontModel.java:
```diff
@@ -19,6 +21,22 @@ public FontModel(@ColorInt int color, String glyph, float textSize, Typeface typ
     }
+    
+    @Override
+    public boolean equals(Object o) {
```
Thanks for this, moved to a new PR.

mattleibow commented:
/azp run

albyrock87 commented:
Closed for #24023
Will open a new PR simply to add some logging around Glide.
